GMEngine/GMEngine/CModuleRender.cpp

```diff
@@ -0,0 +1,171 @@
+	static float light3 = 1.f;
+	bool b = true;
+
+	for (std::multimap<GLuint, renderData>::iterator it = renderQueue.begin(); it != renderQueue.end(); ++it)
+	{
+		if ((*it).first != program)
+		{
+			program = (*it).first;
+			glUseProgram(program);
+		}
+
+		GLint modelLoc = glGetUniformLocation(program, "model_matrix");
+		glUniformMatrix4fv(modelLoc, 1, GL_FALSE, &(*it).second.id[0][0]);
+
+		GLint viewLoc = glGetUniformLocation(program, "view");
+		//glUniformMatrix4fv(viewLoc, 1, GL_FALSE, App->camera->getViewMatrix());
+
+		GLint projectLoc = glGetUniformLocation(program, "projection");
+		//glUniformMatrix4fv(projectLoc, 1, GL_FALSE, App->camera->getProjectMatrix());
+
+		GLint normalLoc = glGetUniformLocation(program, "normal_matrix");
+		if (normalLoc != -1)
+		{
+			glUniformMatrix4fv(normalLoc, 1, GL_FALSE, &(*it).second.normalId[0][0]);
+		}
+
+		GLint lightPosLoc = glGetUniformLocation(program, "lightPosition");
+		if (lightPosLoc != -1)
+		{glUniform3f
+			(lightPosLoc, light1, light2, light3);
+		}
+
+		//float3 cameraPosition = App->camera->currentCamera->frustum.pos;
+
+		GLint cameraPosLoc = glGetUniformLocation(program, "cameraPosition");
+		if (cameraPosLoc != -1)
+		{
+			//glUniform3f(cameraPosLoc, cameraPosition.x, cameraPosition.y, cameraPosition.z);
+		}
+
+		glUniform1i(glGetUniformLocation(program, "useText"), (*it).second.hasTexture);
+		glUniform1i(glGetUniformLocation(program, "useNormal"), (*it).second.hasNormalMap);
+
+		//Order matters!
+		if ((*it).second.hasTexture)
+		{
+			glActiveTexture(GL_TEXTURE0);
+			glBindTexture(GL_TEXTURE_2D, (*it).second.textureID);
+			glUniform1i(glGetUniformLocation(program, "text"), 0);
+
+			glBindBuffer(GL_ARRAY_BUFFER, (*it).second.textureCoordsID);
+			glEnableVertexAttribArray(1);
+			glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 2 * sizeof(GLfloat), (GLvoid*)0);
+
+			if ((*it).second.hasNormalMap)
+			{
+				glActiveTexture(GL_TEXTURE1);
+				glBindTexture(GL_TEXTURE_2D, (*it).second.normalMapID);
+				glUniform1i(glGetUniformLocation(program, "normalMap"), 0);
+
+				glBindBuffer(GL_ARRAY_BUFFER, (*it).second.textureCoordsID);
+				glEnableVertexAttribArray(1);
+				glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 2 * sizeof(GLfloat), (GLvoid*)0);
+			}
+		}
+
+		glBindBuffer(GL_ARRAY_BUFFER, (*it).second.idVertVBO);
+		glEnableVertexAttribArray(0);
+		glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(GLfloat), (GLvoid*)0);
+
+		glBindBuffer(GL_ARRAY_BUFFER, (*it).second.colorID);
+		glEnableVertexAttribArray(2);
+		glVertexAttribPointer(2, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(GLfloat), (GLvoid*)0);
+
+		glBindBuffer(GL_ARRAY_BUFFER, (*it).second.idNormalVBO);
+		glEnableVertexAttribArray(3);
+		glVertexAttribPointer(3, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(GLfloat), (GLvoid*)0);
+
+		glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, (*it).second.idIdxVAO);
+
+		glDrawElements((*it).second.mode, (*it).second.sizeIdxVAO, GL_UNSIGNED_INT, NULL);
+	}
+
+	renderQueue.clear();
+}
```

Comment:
We should add Camera component soon to remove this comments, or remove the comments.